Cross-browser extension article update (#30228)
mdn/content: files/en-us/mozilla/add-ons/webextensions/build_a_cross_browser_extension/index.md

{{AddonSidebar()}}

The introduction of the browser extensions API created a uniform landscape for the development of browser extensions. However, there are differences in the API implementations and the scope of coverage among the browsers that use the extensions API (the major ones being Chrome, Edge, Firefox, Opera, and Safari).

Maximizing the reach of your browser extension means developing it for at least two browsers, possibly more. This article looks at the main challenges faced when creating a cross-browser extension and suggests how to address these challenges.

> **Note:** The main browsers have adopted Manifest V3. This manifest version provides better compatibility between the browser extension environments, such as promises for handling asynchronous events. In addition to the information in this guide, refer to the Manifest V3 migration guides for [Firefox](https://extensionworkshop.com/documentation/develop/manifest-v3-migration-guide/) and [Chrome](https://developer.chrome.com/docs/extensions/migrating/).

## Cross-platform extension coding hurdles

> There are eight areas you need to address when tackling a cross-platform extension:

Author: Point taken, although it's a shame, as the count does give the reader an impression of scope of activities

- API namespace
- API asynchronous event handling
- API function coverage

> - Content script execution contexts

Reviewer: Each bullet point is mentioned in this article. Could you turn each bullet point in a link for ease of navigation?

> - Background scripts, page, and service worker

Author: @Rob--W any thoughts/comments?

> - Background scripts, page, and service worker

Reviewer 2: Looks good to me.

Firefox also supports the `chrome.*` namespace for APIs that are compatible with Chrome, primarily to assist with [porting](https://extensionworkshop.com/documentation/develop/porting-a-google-chrome-extension/). However, using the `browser.*` namespace is preferred. In addition to being the proposed standard, `browser.*` uses promises—a modern and convenient mechanism for handling asynchronous events.

Only in the most trivial extensions is namespace likely to be the only cross-platform issue to be addressed. Therefore, it's rarely, if ever, helpful to address this issue alone. The best approach is to address this with asynchronous event handling.

### API asynchronous event handling

> With the introduction of Manifest V3 (Chrome 88), all the main browsers support _Promises_ as their approach to handling asynchronous events. Before that, Firefox and Safari used _Promises_, while _Callbacks_ were used by Chrome, Edge, and Opera.

Reviewer: Suggested edit, to clarify more precisely when callbacks/promises are supported, and reference more documentation/examples. Additionally, I also added a paragraph to explain the differences in handling events asynchronously, because the title of this section is "API asynchronous event handling", yet there was no explanation of "event handling" at all.

```suggestion
All main browsers return _Promises_ from asynchronous methods in Manifest V3. In Manifest V2, Chrome's asynchronous methods do not return Promises, but invoke _callbacks_ instead. For compatibility, all main browsers support callbacks across all manifest versions. See [Callbacks and Promises](/en-US/docs/Mozilla/Add-ons/WebExtensions/Chrome_incompatibilities#callbacks_and_promises) for details.

Some handlers of extension API events are expected to respond asynchronously, through a `Promise` or callback function. For example, a handler of the `runtime.onMessage` event can [send an asynchronous response using a `Promise`](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage#sending_an_asynchronous_response_using_a_promise) or using [a callback](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage#sending_an_asynchronous_response_using_sendresponse). A `Promise` as return value from an event handler is supported in Firefox and Safari, but not yet in Chrome.
```

Reviewer 2: Line 41 is problematic for Chrome since these were implemented over the span of 88 to 120 (maybe 121). I'm working on a suggestion for what to say here. Ping me directly if this blocks the merge.

Reviewer: Line 41 of the original only, or also my suggested edit?

Author: @jpmedley This is part of work to catch up on the backlog, so merging isn't urgent. However, obviously, it would be nice to get it done in a timely manner

Reviewer 2: Here's the gist. Word it as you see fit:
* In Chrome Manifest V3 did not ship with support for many promises. 
* Promises have been gradually added since then.
* See individual Chrome APIs to learn when promise support was added.
[Here's an example](https://developer.chrome.com/docs/extensions/reference/cookies/#method-getAll:~:text=cookie%20was%20found.-,RETURNS,-Promise%3CCookie) so you can see how we render this.

They tell me it will be wrapped in 120 or 121. (We're at 119 now.) I won't really know for sure until I see it. We don't have a tracking bug for this.

Author: How about:
```suggestion
With the introduction of Manifest V3, all the main browsers adopted the standard of returning _Promises_ from asynchronous methods. However, at the time of writing, November 2023, Chrome needed to implement Promises on some APIs. The implementation of Promises is expected to be complete by Chrome 121. Consult the Chrome documentation for details of whether the APIs you use have Promise support.

In Manifest V2, Firefox and Safari support Promises for asynchronous methods. At the same time, Chrome methods invoke _callbacks_. For compatibility, all the main browsers support callbacks across all manifest versions. See [Callbacks and Promises](/en-US/docs/Mozilla/Add-ons/WebExtensions/Chrome_incompatibilities#callbacks_and_promises) for details.

Some handlers of extension API events are expected to respond asynchronously through a `Promise` or callback function. For example, a handler of the `runtime.onMessage` event can [send an asynchronous response using a `Promise`](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage#sending_an_asynchronous_response_using_a_promise) or using [a callback](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage#sending_an_asynchronous_response_using_sendresponse). A `Promise` as the return value from an event handler is supported in Firefox and Safari, but not yet in Chrome.
```

Reviewer 2: Looks good to me.

> > **Note:** As of November 2023, the Chrome documentation notes that "support has been added to most methods, though callbacks are still supported as an alternative. (We will eventually support promises on all appropriate methods.)"

Author: Thanks @jpmedley I'll await your update. Has a bit of a knock-on impact into other work too so would be good to know.

> > **Note:** As of November 2023, the Chrome documentation notes that "support has been added to most methods, though callbacks are still supported as an alternative. (We will eventually support promises on all appropriate methods.)"

Reviewer: Let's drop this note. I have suggested an edit to clarify the callback support, and the "(We will eventually support promises on all appropriate methods.)" remark is unnecessary.

The vast majority of Chrome extension APIs support Promises now.

> > **Note:** As of November 2023, the Chrome documentation notes that "support has been added to most methods, though callbacks are still supported as an alternative. (We will eventually support promises on all appropriate methods.)"

Reviewer 2: Dropping this seems right. Chrome's promise work will be done this quarter. The phrase 'all appropriate methods' is right. There are places where callbacks are appropriate and will remain in use.

> Firefox also supports callbacks for the APIs that support the `chrome.*` namespace. However, using promises (and the `browser.*` namespace) is recommended. Promises greatly simplifies asynchronous event handling, particularly where you need to chain events together.

Author: @jpmedley, how about this?
```suggestion
Firefox also supports callbacks for the APIs that support the `chrome.*` namespace. However, using promises is recommended. Promises greatly simplifies asynchronous event handling, particularly where you need to chain events together. This means using a polyfill or similar so your extension uses the `browser.*` namespace in Firefox and Safari and `chrome.*` in Chrome, Opera, and Edge.
```

> Firefox also supports callbacks for the APIs that support the `chrome.*` namespace. However, using promises (and the `browser.*` namespace) is recommended. Promises greatly simplifies asynchronous event handling, particularly where you need to chain events together.

Reviewer: FYI: Chrome is supportive of a `browser` namespace, see https://bugs.chromium.org/p/chromium/issues/detail?id=798169#c10 for details. May be worth linking here for those who are interested in following the progress?

> Firefox also supports callbacks for the APIs that support the `chrome.*` namespace. However, using promises (and the `browser.*` namespace) is recommended. Promises greatly simplifies asynchronous event handling, particularly where you need to chain events together.

Reviewer 2: What he says is true, but I didn't mention it because I have no idea when we'll get to it. 

I oppose adding bug links to documentation. They imply promises that may or may not be fulfilled and they tend to hang around long after their needed.

> Firefox also supports callbacks for the APIs that support the `chrome.*` namespace. However, using promises (and the `browser.*` namespace) is recommended. Promises greatly simplifies asynchronous event handling, particularly where you need to chain events together.

Reviewer: U think that it's useful to communicate future support for the browser namespace in Chrome, including the bug number. Then extension authors can reference a bug number in their source code.

And also to avoid code like `const IS_FIREFOX = typeof browser == "object";`

> Firefox also supports callbacks for the APIs that support the `chrome.*` namespace. However, using promises (and the `browser.*` namespace) is recommended. Promises greatly simplifies asynchronous event handling, particularly where you need to chain events together.

Author: @jpmedley it's not been unusual for the MDN documentation to include links to bugs. That said @Rob--W, while I haven't done an exhaustive analysis, most of those links to my recollection relate to bugs or similar things where the documentation might get missed. When Chrome makes this change isn't this likely to be fairly big news and unlikely to be missed? But also, until then knowing that it's a possibility will have no effect on how people develop cross-browser extensions.

> Firefox also supports callbacks for the APIs that support the `chrome.*` namespace. However, using promises (and the `browser.*` namespace) is recommended. Promises greatly simplifies asynchronous event handling, particularly where you need to chain events together.

Reviewer 2: I have to assume it will be big news also.

> Firefox also supports callbacks for the APIs that support the `chrome.*` namespace. However, using promises (and the `browser.*` namespace) is recommended. Promises greatly simplifies asynchronous event handling, particularly where you need to chain events together.

Reviewer 2: The addition sounds good.

Firefox also supports the `chrome.*` namespace for APIs that are compatible with Chrome, primarily to assist with [porting](https://extensionworkshop.com/documentation/develop/porting-a-google-chrome-extension/). However, using the `browser.*` namespace is preferred. In addition to being the proposed standard, `browser.*` uses promises—a modern and convenient mechanism for handling asynchronous events.

Only in the most trivial extensions is namespace likely to be the only cross-platform issue to be addressed. Therefore, it's rarely, if ever, helpful to address this issue alone. The best approach is to address this with asynchronous event handling.

### API asynchronous event handling

> With the introduction of Manifest V3, all the main browsers adopted the standard of returning _Promises_ from asynchronous methods. However, at the time of writing, November 2023, Chrome needed to implement Promises on some APIs. The implementation of Promises is expected to be complete by Chrome 121. Consult the Chrome documentation for details of whether the APIs you use have Promise support.

Reviewer: Chrome 121 is in a feature freeze, with the branch cut being scheduled for 4 December. The crbugs tracking Promise support are still open. I think that the statement here is too committal and that it's better to rephrase it, as suggested below. An alternative to the vague wording is to take a full inventory of which APIs support async methods and then be explicit. If we are not in a rush I can check this myself in the source code (or consult a Chrome extension team member to do that).

```suggestion
With the introduction of Manifest V3, all the main browsers adopted the standard of returning _Promises_ from asynchronous methods. Firefox and Safari have full support for Promises on all asynchronous APIs. At the time of writing, November 2023, Chrome has nearly complete support for Promises on asynchronous APIs. Consult the Chrome documentation for details of whether the APIs you use have Promise support.
```

Author: @Rob--W I'll leave it to your call on whether to add a full inventory. However, is it worth it for a list that could quickly go out of date? If you do a list, would it be best for the API not done? I get the impression that might be the shorter list.

> With the introduction of Manifest V3, all the main browsers adopted the standard of returning _Promises_ from asynchronous methods. Firefox and Safari have full support for Promises on all asynchronous APIs. At the time of writing, November 2023, Chrome has nearly complete support for Promises on asynchronous APIs. Consult the Chrome documentation for details of whether the APIs you use have Promise support.

Reviewer: Since the PR has started, engineering work has completed in Chrome 121 that makes Promise support opt-out rather than opt in (https://chromium.googlesource.com/chromium/src/+/4d69e3840fbb82b93651c2be152d9bba5c453bbf). Because of that, I suggest to remove the vague time aspect of the statement here, in favor of a precise statement. Chrome 121 is currently on Beta FYI.

```suggestion
With the introduction of Manifest V3, all the main browsers adopted the standard of returning _Promises_ from asynchronous methods. Firefox and Safari have full support for Promises on all asynchronous APIs. Starting from Chrome 121, all asynchronous extension APIs support promises unless documented otherwise. The `devtools` API is the only API namespace without Promise support ([Chromium bug 1510416](https://crbug.com/1510416)).
```

}
```

### Content script execution contexts

Content scripts can access and modify the page's DOM, just as page scripts can. They can also see any changes made to the DOM by page scripts. However, content scripts get a "clean" view of the DOM.

Firefox and Chrome use fundamentally different approaches to handle this behavior: in Firefox it's called Xray vision, while Chrome uses isolated worlds. For more details, see [Content script environment](/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts#content_script_environment) section of the Content scripts concept article.

However, Firefox provides some APIs that enable content scripts to access JavaScript objects created by page scripts and to expose their JavaScript objects to page scripts. See [Sharing objects with page scripts](/en-US/docs/Mozilla/Add-ons/WebExtensions/Sharing_objects_with_page_scripts) for details.

> 

Reviewer: Let's also mention the difference in CSP in the content script. Here is a suggested minimal sentence. It may be useful to also put this sentence in the Content_scripts article, since it's quite relevant in some cases.

```suggestion

There are also differences between the [Content Security Policy (CSP) for content scripts](/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_Security_Policy#csp_for_content_scripts).
```

Reviewer 2: +1

> ### Background scripts, page, and service worker

Reviewer 2: +1

> As part of its implementation of Manifest V3, Chrome replaced background pages with extension service workers. Firefox retains the use of background pages and scripts, while Safari supports background pages and scripts and service workers.

Reviewer: The `scripts` key is just a convenient way to declare the scripts that are loaded in an auto-generated background page. Let's just mention "background page" instead of "background page and scripts".

```suggestion
As part of its implementation of Manifest V3, Chrome replaced background pages with extension service workers. Firefox retains the use of background pages, while Safari supports background pages and service workers.
```

Firefox and Chrome use fundamentally different approaches to handle this behavior: in Firefox it's called Xray vision, while Chrome uses isolated worlds. For more details, see [Content script environment](/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts#content_script_environment) section of the Content scripts concept article.

However, Firefox provides some APIs that enable content scripts to access JavaScript objects created by page scripts and to expose their JavaScript objects to page scripts. See [Sharing objects with page scripts](/en-US/docs/Mozilla/Add-ons/WebExtensions/Sharing_objects_with_page_scripts) for details.

There are also differences between the [Content Security Policy (CSP) for content scripts](/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_Security_Policy#csp_for_content_scripts).

### Background page and extension service worker

As part of its implementation of Manifest V3, Chrome replaced background pages with extension service workers. Firefox retains the use of background pages, while Safari supports background pages and service workers.

> For more information, see the [browser support](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/background#browser_support) section on the `"background"`` manifest key page. This includes a simple example of how to implement a cross-browser script.

Reviewer: Double backtick should be a single backtick.

```suggestion
For more information, see the [browser support](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/background#browser_support) section on the `"background"` manifest key page. This includes a simple example of how to implement a cross-browser script.
```

## Conclusion

When approaching a cross-platform extension development, the differences between extension API implementations can be addressed by targeting Firefox and using the [WebExtension browser API Polyfill](https://github.com/mozilla/webextension-polyfill/).

> The bulk of your cross-platform work is likely to focus on handling variations among the API features supported by the main browsers. You may also need to account for differences between the content and background script implementations. Creating your `manifest.json` files should be relatively straightforward and something you can do manually. You then need to account for the variations in the processes for submitting to each extension store.

Reviewer: ```suggestion
The bulk of your cross-platform work is likely to focus on handling variations among the API features supported by the main browsers. You may also need to account for differences between the content script and background script implementations. Creating your `manifest.json` files should be relatively straightforward and something you can do manually. You then need to account for the variations in the processes for submitting to each extension store.
```
